### PR TITLE
feat: Add pluggable services architecture in lfx and comprehensive testing

### PR DESCRIPTION
**Pull request by ogabrielluiz**
Introduce a pluggable services architecture with detailed documentation and an example configuration file. Add unit and integration tests for various services, including LocalStorageService, TelemetryService, and VariableService, ensuring robust coverage and error handling. Implement service registration via decorators to enhance service discovery and management.

## Summary by CodeRabbit

- New Features
  - Introduces a pluggable services system with discovery via configuration, decorators, and entry points.
  - Adds minimal built-in services: Telemetry, Tracing, and Variable (in-memory with env fallback).
  - Enhances Storage service lifecycle with teardown support.
  - Supports configuration via lfx.toml or pyproject.toml for service overrides.

- Documentation
  - New guide detailing pluggable services, configuration, discovery, and troubleshooting.
  - README updates: Pluggable Services overview and flattened component access.
  - Adds lfx.toml example for common service mappings.

- Tests
  - Comprehensive unit and integration tests covering discovery, overrides, DI, lifecycle, and edge cases.

### CODE REVIEW

**Comment by ogabrielluiz** [comment]
@coderabbitai review

**Comment by mpawlow** [comment]
### Status Update

- ✅ Rebased the branch off of the latest from master
- ✅ Resolved all conflicts
- ✅ Added fix: Remove duplicate teardown method from LocalStorageService
- ✅ Force pushed back to remote
- ✅ Code Quality Checks - PASSED
   - ✅ make format_backend - All checks passed
   - ✅ make lint - No issues found (mypy clean)
   - ✅ No conflict markers remaining
   - ✅ No whitespace errors
   - ✅ All pre-commit hooks passed
- ✅ Updated service tests for LocalStorageService constructor changes
   - ✅ Tests: 154/154 passing

### Test Failures - Full Details

- 23 test failures in src/lfx/tests/unit/services/ related to LocalStorageService constructor signature changes:
   - Tests are using old constructor: LocalStorageService(data_dir=tmp_path)
   - New constructor requires: LocalStorageService(session_service, settings_service)
1. Fixed Direct Instantiation Tests (test_minimal_services.py)
   - Created mock_session_service and mock_settings_service fixtures
   - Updated storage fixture to use: LocalStorageService(mock_session_service, mock_settings_service)
   - Updated all integration tests to use the new fixtures
2. Fixed ServiceManager Tests (3 files)
   - Created MockSessionService class in each test file
   - Registered MockSessionService in ServiceManager fixtures so dependency resolution works
   - Updated service count assertions to account for the mock service
   - Files updated:
      - test_service_manager.py
      - test_decorator_registration.py
      - test_integration.py
      - test_edge_cases.py

**Comment by mpawlow** [comment]
@coderabbitai configuration

**Comment by mpawlow** [comment]
@coderabbitai review

File: src/lfx/src/lfx/services/manager.py [hunk 1]
```diff
@@ -75,6 +121,117 @@ def get(self, service_name: ServiceType, default: ServiceFactory | None = None)
...
+                    raise RuntimeError(msg)
+                # Recursively create dependency if not exists
+                # Note: Thread safety is handled by the caller's keyed lock context
+                if dependency_type not in self.services:
+                    self._create_service(dependency_type)
+                dependencies[param_name] = self.services[dependency_type]
+
+        # Create the service instance
+        try:
+            service_instance = service_class(**dependencies)
```
Comment (HimavarshaVS) [inline src/lfx/src/lfx/services/manager.py:188]: Seem like we are storing the partial state even before the validation is complete. 
can do something like 
```suggestion
            service_instance = service_class(**dependencies)
            # Validate service has required attributes
            if not hasattr(service_instance, 'name'):
                raise ValueError(f"Service {service_class.__name__} missing 'name' attribute")
                -----
```

File: src/lfx/src/lfx/services/manager.py [hunk 2]
```diff
@@ -75,6 +121,117 @@ def get(self, service_name: ServiceType, default: ServiceFactory | None = None)
...
         """Create a new service given its name, handling dependencies."""
         logger.debug(f"Create service {service_name}")
+
+        # Settings service is special - always use factory, never from plugins
+        if service_name == ServiceType.SETTINGS_SERVICE:
+            self._create_service_from_factory(service_name, default)
+            return
+
+        # Try plugin discovery first (if not already done)
+        if not self._plugins_discovered:
```
Comment (HimavarshaVS) [inline src/lfx/src/lfx/services/manager.py:131]:
```suggestion
       with self._lock:
            if not self._plugins_discovered:
                # ... discovery logic ...
                self._plugins_discovered = True
```
Suggesting this as _plugins_discovered is set without synchronization which can lead to triggering multiple discoveries.

**Review by HimavarshaVS — APPROVED** [review]
few comments to avoid race conditions . But other than that lgtm

File: src/lfx/src/lfx/services/manager.py (hunk 2 above)
Comment (mpawlow) [reply src/lfx/src/lfx/services/manager.py:131]: - Excellent catch.
- This is indeed a valid problem
- Added an additional commit to address this issue

File: src/lfx/src/lfx/services/manager.py (hunk 1 above)
Comment (mpawlow) [reply src/lfx/src/lfx/services/manager.py:188]: - I need to circle back and evaluate this further before making any destabilizing changes
- Not sure if validating properties is sufficient or explicit validation should be done via introducing a `validate` lifecycle method or deferring validation to the `set_ready` method

File: src/lfx/src/lfx/services/manager.py [hunk 3]
```diff
@@ -75,6 +121,117 @@ def get(self, service_name: ServiceType, default: ServiceFactory | None = None)
...
+                        pass
+                    continue
+
+            if dependency_type:
+                # Check for circular dependency (service depending on itself)
+                if dependency_type == service_name:
+                    msg = f"Circular dependency detected: {service_name.value} depends on itself"
+                    raise RuntimeError(msg)
+                # Recursively create dependency if not exists
+                # Note: Thread safety is handled by the caller's keyed lock context
```
Comment (jordanrfrazier) [inline src/lfx/src/lfx/services/manager.py]: Does this potential race condition exist? 
```
Thread A: 
* get(serviceA) -> lock(serviceA) -> dependsOn(ServiceC) -> _create_service(serviceC) (no serviceC locking)
Thread B:
* get(serviceB) -> lock(serviceB) -> dependsOn(ServiceC) -> _create_service(serviceC) (no serviceC locking)
```

If so, we can do something like:
```
+                with self.keyed_lock.lock(dependency_type):
+                    if dependency_type not in self.services:
+                        self._create_service(dependency_type)
+                    dependencies[param_name] = self.services[dependency_type]

```

File: src/lfx/src/lfx/services/manager.py [hunk 4]
```diff
@@ -75,6 +121,117 @@ def get(self, service_name: ServiceType, default: ServiceFactory | None = None)
...
+                    if param.default == inspect.Parameter.empty:
+                        # No default, can't resolve - will fail during instantiation
+                        pass
+                    continue
+
+            if dependency_type:
+                # Check for circular dependency (service depending on itself)
+                if dependency_type == service_name:
+                    msg = f"Circular dependency detected: {service_name.value} depends on itself"
+                    raise RuntimeError(msg)
```
Comment (jordanrfrazier) [inline src/lfx/src/lfx/services/manager.py]: This does only catch A -> A dependencies, I don't believe it will catch A -> B -> A. The error message does concur with that statement, but we may want to add logic to catch the latter scenarios.
